logger.go

```diff
@@ -0,0 +1,175 @@
+package glog
+
+import (
+	"context"
+	"fmt"
+)
+
+// Logger provides logging functionality with additional data and prefixing.
+type Logger struct {
+	*loggingT
+	prefix string
+	// Extra arguments to be appended to each request
+	data []interface{}
+}
+
+// NewLogger creates a Logger instance with no additional data.
+func NewLogger() Logger {
+	return Logger{
+		loggingT: &logging,
+	}
+}
+
+// WithContext creates a logger from a context.Context
+func WithContext(ctx context.Context) Logger {
+	data, _ := ctx.Value(contextKeyData).([]interface{})
+	prefix, _ := ctx.Value(contextKeyPrefix).(string)
+	return Logger{
+		loggingT: &logging,
+		data:     data,
+		prefix:   prefix,
+	}
+}
+
+// WithPrefix creates a Logger with a given prefix.
+func WithPrefix(prefix string) Logger {
+	return Logger{
+		loggingT: &logging,
+		prefix:   prefix,
+	}
+}
+
+// WithData creates a Logger with a given set of backend data.
+// All args provided will be wrapped with glog.Data prior to being sent.
+func WithData(args ...interface{}) Logger {
+	return Logger{
+		loggingT: &logging,
+		data:     args,
+	}
+}
+
+// WithPrefix creates a Logger from an existing logger with a specified prefix.
+// Any prefix on the input Logger will be replaced.
+func (l Logger) WithPrefix(prefix string) Logger {
+	newLogger := l
+	newLogger.prefix = prefix
+	return newLogger
+}
+
+// WithData creates a Logger from an existing logger, using the specifed data.
+// Any existing data on the input Logger will be replaced.
+func (l Logger) WithData(vars ...interface{}) Logger {
```

Comment:
Fro the sake of syntactic sugar, I would also add WithPrefixAndData. Context is not always available and building it just for the sake of logging seems wasteful.

Reply:
Makes sense :)

Reply:
Actually, had a think about this, and I don't think we should add a function to combine two existing ones. This leaves us with two ways to do the same thing:
- WithData(...).WithPrefix(...)
- WithDataAndPrefix(...,...)
Which would invite confusion.

We should think of this as a builder, otherwise we'll have to add various combinations of "and" functions every time we add something new that can be added to a Logger.

Note that Context.WithValue does not allow adding multiple values at once, and nobody worries about the copying there.

Reply:
Makes perfect sense. I forgot we can chain them :)
I still think it should be a pointer, though ;)